### PR TITLE
:computer: :art: Output vector data of particles for viz

### PR DESCRIPTION
**Pull request by kks32**
This should define displacements as a new particle vector field, overrides #472

### CODE REVIEW

File: include/particle.tcc [hunk 1]
```diff
@@ -575,6 +576,8 @@ bool mpm::Particle<Tdim, Tnphases>::compute_updated_position(unsigned phase,
 
       // New position  current position + velocity * dt
       this->coordinates_ += nodal_velocity * dt;
+      // Update displacement (displacement is initialized from zero)
+      this->displacement_ += nodal_velocity * dt;
```
Comment (ezrayst) [inline include/particle.tcc:590]: I like this!

File: include/particle.tcc [hunk 2]
```diff
@@ -719,3 +724,20 @@ void mpm::Particle<Tdim, Tnphases>::apply_particle_velocity_constraints() {
     this->velocity_(direction, phase) = constraint.second;
   }
 }
+
+//! Apply particle velocity constraints
```
Comment (ezrayst) [inline include/particle.tcc]: This comment does not fit the function

File: include/particle.tcc [hunk 3]
```diff
@@ -719,3 +724,20 @@ void mpm::Particle<Tdim, Tnphases>::apply_particle_velocity_constraints() {
...
+template <unsigned Tdim, unsigned Tnphases>
+Eigen::VectorXd mpm::Particle<Tdim, Tnphases>::vector_data(
+    unsigned phase, const std::string& property) {
+  //! Map of properties
+  std::map<std::string, std::function<Eigen::VectorXd(unsigned)>> properties;
```
Comment (ezrayst) [inline include/particle.tcc]: I wonder whether we should make `map` every time we call `vector_data`. My thoughts would be to have the map as private variables.

**Review by ezrayst — COMMENTED** [review]
Like the general structure!
When one day we output `state variables`, how should we expand this?

File: include/particle.tcc (hunk 3 above)
Comment (kks32) [reply include/particle.tcc]: the problem is we can't pass phase otherwise? I'm not quite sure about this.

Comment (ezrayst) [reply include/particle.tcc]: I see. How about a map that has two keys (attribute and phase), like a tuple? We can define the phase up to 3 only right, not infinite.

I still feel making map for every particle and every time step we output is a bit inefficient.

**Review by ezrayst — APPROVED** [review]

**Pull request merged**